lib/services/access_request.go

```diff
@@ -245,53 +344,212 @@ func NewAccessRequest(user string, roles ...string) (AccessRequest, error) {
 	return &req, nil
 }
 
+func (c AccessRequestConditions) GetTraitMappings() TraitMappingSet {
+	tm := make([]TraitMapping, 0, len(c.ClaimsToRoles))
+	for _, mapping := range c.ClaimsToRoles {
+		tm = append(tm, TraitMapping{
+			Trait: mapping.Claim,
+			Value: mapping.Value,
+			Roles: mapping.Roles,
+		})
+	}
+	return TraitMappingSet(tm)
+}
+
 type UserAndRoleGetter interface {
 	UserGetter
 	RoleGetter
+	GetRoles() ([]Role, error)
 }
 
-func ValidateAccessRequest(getter UserAndRoleGetter, req AccessRequest) error {
-	user, err := getter.GetUser(req.GetUser(), false)
+// requestValidator collects state related to
+// access request validation.
+type requestValidator struct {
+	traits map[string][]string
+	state  RequestState
+	Roles  struct {
+		Allow, Deny []parse.Matcher
+	}
+	Annotations struct {
+		Allow, Deny map[string][]string
+	}
+}
+
+func newRequestValidator(traits map[string][]string, state RequestState) requestValidator {
+	m := requestValidator{
+		traits: traits,
+		state:  state,
+	}
+	if m.state.IsPending() {
+		// if request state is PENDING, we need to build
+		// system annotations.
+		m.Annotations.Allow = make(map[string][]string)
+		m.Annotations.Deny = make(map[string][]string)
+	}
+	return m
+}
+
+// push integrates a role's configuration into the request validator.
+// All roles must be pushed before validation begins.
+func (m *requestValidator) push(role Role) error {
```

Comment:
This entire function is pretty difficult to understand what it's doing. Can it be refactored to break up the logic to make it simpler?